src/util/treeToFlatArray.js

```diff
@@ -0,0 +1,243 @@
+  const childPath = unroot(parentPath.concat(child))
+  return ({
+    hiddenNodes: acc.hiddenNodes + (pathToContext(getThoughtsRanked(childPath)).includes('=hidden') ? 1 : 0)
+  })
+}, {
+  hiddenNodes: 0
+})
+
+const getFirstSubthoughtValue = path => {
+  const children = getThoughts(path)
+  return children.length > 0 ? children[0].value : null
+}
+
+// recursively finds all the visible thought and returns a flat array
+const getFlatArray = ({
+  startingPath,
+  cursor,
+  children,
+  isLeaf,
+  showHiddenThoughts,
+  isParentCursorAncestor = true,
+  isCursorDescendant = false,
+  visibleSiblingsCount,
+  pinChildren,
+  viewInfo = { table: { tableFirstColumnsAbove: 0, tableSecondColumnsAbove: 0, isActive: false, column: null } }
+} = {}) => {
+  const parentNode = head(startingPath) || RANKED_ROOT[0]
+
+  const subThoughts = children || getThoughtsRanked(startingPath).filter(child => showHiddenThoughts || !isFunction(child.value))
+  const isCursorContext = equalPath(startingPath, contextOf(cursor))
+
+  // iterate subthoughts
+  return subThoughts.reduce((acc, child, index) => {
+    const childPath = unroot(startingPath.concat(child))
+    const value = child.value
+
+    // isParentCursorAncestor is used to prevent calling isDescendant everytime
+    // if the parent thought is already not an ancestor of the cursor then we don't need to call it everytime for its descendants
+    const isCursorAncestor =
+      isParentCursorAncestor &&
+      isDescendant(pathToContext(childPath), pathToContext(cursor))
+    const isCursor = equalPath(cursor, childPath)
+    const childPathLength = childPath.length
+
+    const children = getThoughtsRanked(childPath)
+
+    // decide if it is a distant ancestor that needs to be visible but needs to stop deeper recursion
+    const addDistantAncestorAndStop =
+      cursor.length - childPathLength <= (isLeaf ? 1 : 0) &&
+      !isCursor &&
+      !isCursorAncestor &&
+      !isCursorDescendant
+    // stop recursion if distant ancestor doesn't need to be added to the array
+    const showDistantAncestor = !(
+      childPathLength < cursor.length &&
+      !isCursorAncestor &&
+      !addDistantAncestorAndStop
+    )
+
+    const { isHidden, isPinned, isChildrenPinned, view, filteredChildren } = children.reduce((acc, child) => {
+      return {
+        isHidden: acc.isHidden || child.value === '=hidden',
+        isPinned: acc.isPinned || child.value === '=pin',
+        isChildrenPinned: acc.isChildrenPinned || child.value === '=pinChildren',
+        view: child.value === '=view' ? getFirstSubthoughtValue(unroot(childPath.concat(child))) : acc.view,
+        filteredChildren: acc.filteredChildren.concat(
+          !showHiddenThoughts && isFunction(child.value) ? [] : [child]
+        )
+      }
+    }, {
+      isHidden: false,
+      isPinned: false,
+      isChildrenPinned: false,
+      filteredChildren: []
+    })
```

Comment:
> 3\. (Hypothetical) Is there any way we can get rid of `filteredChildren` completely? It would be nice if we could let the next recursive iteration calculate `children` on its own and not have to pass it through. The only thing it's being used for is `visibleSiblingsCount`, but I see that is part of the `stop` functionality so it's not obvious to me if this would even be possible.

`filteredChildren` is used to calculate `visibleSiblingsCount` which is passed to children nodes. Since the recursive implementation is depth first , so the children would always need to iterate all it's siblings to calculate the visible siblings. If we pass them the total count of visible nodes at their depth from the parent itself ,then they won't have to do that calculation.

We discussed about filtering children's here.
https://github.com/cybersemics/em/pull/503#discussion_r413457498

Reply:
> Access attributes directly:
> 
> * `isHidden`
> * `isPinned`
> * `isChildrenPinned`
> * `view`
> 
> `attributeEquals` and `attribute(..., '=view') are both `O(1)`since they only have one child.`=hidden`is the only attribute that behaves differently, but I believe it was a mistake and should have a value of`true`or`false`for performance reasons. You are welcome to use`hasAttribute`with`=hidden` for now and we can change it in a separate task.

Yes it makes perfect sense to do this. I will make these changes!